bionic/dagviz.py

```diff
@@ -91,14 +91,39 @@ def dot_from_graph(graph, vertical=False, curvy_lines=False, name=None):
         rankdir="TB" if vertical else "LR",
     )
 
-    node_lists_by_cluster = defaultdict(list)
+    # First, we cluster together any nodes that share an entity name. This includes
+    # nodes generated by a common tuple function, and nodes that differ only by case
+    # key.
+    node_clusters_by_entity_name = defaultdict(list)
     for node in graph.nodes():
-        descriptor = graph.nodes[node]["descriptor"]
-        node_lists_by_cluster[descriptor].append(node)
-
-    descriptors = list(set(graph.nodes[node]["descriptor"] for node in graph.nodes()))
-    color_strs_by_descriptor = hpluv_color_dict(
-        descriptors, saturation=99, lightness=90
+        entity_names = list(node.dnode.all_entity_names())
+        # If this node has no entity names at all, we'll pretend it has a single entity
+        # name. (This is unlikely to happen in practice, but could happen for a
+        # descriptor like the empty tuple `(,)`.)
+        if len(entity_names) == "":
+            entity_names = [""]
+
+        # If this node has multiple entity names, we want to make sure all those names
+        # are associated with the same cluster, so we'll merge all the clusters we find.
+        first_cluster = node_clusters_by_entity_name[entity_names[0]]
+        for other_entity_name in entity_names[1:]:
+            other_cluster = node_clusters_by_entity_name[other_entity_name]
+            if other_cluster is not first_cluster:
+                first_cluster.extend(other_cluster)
+                node_clusters_by_entity_name[other_entity_name] = first_cluster
+
+        first_cluster.append(node)
+
+    # Now we deduplicate the clusters and arrange them in a deterministic order.
+    unique_node_clusters_by_obj_id = {
+        id(cluster): cluster for cluster in node_clusters_by_entity_name.values()
+    }
+    sorted_node_clusters = sorted(unique_node_clusters_by_obj_id.values(), key=min)
```

Comment:
Why are we sorting the clusters now whereas we were not sorting them before?

Reply:
We want the assignment of colors to clusters to be deterministic. Previously we clustered by descriptor and then assigned a color to each descriptor; now we assign each cluster an arbitrary index and then assign colors to the indices. Sorting the clusters first (using each cluster's first descriptor as a key) means we always assign the same index to each cluster.

I'll add a comment to make this clearer.